### PR TITLE
KAA-925 Extract kaa_extension_request_get_size()

### PR DESCRIPTION
**Pull request by rasendubi**

### CODE REVIEW

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 1]
```diff
@@ -114,232 +131,190 @@ kaa_error_t kaa_platform_protocol_create(kaa_platform_protocol_t **platform_prot
...
     }
-    char *extension_count_p = writer->current;
+
+    uint16_t *extension_count_p = (uint16_t *)writer->current;
     writer->current += KAA_PROTOCOL_EXTENSIONS_COUNT_SIZE;
+    // TODO: assert KAA_PROTOCOL_EXTENSIONS_COUNT_SIZE == sizeof(uint16_t)
 
     error_code = kaa_meta_data_request_serialize(self, writer, self->request_id);
+    if (error_code) {
+        return error_code;
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: `writer` has leaked.

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 2]
```diff
@@ -114,232 +131,190 @@ kaa_error_t kaa_platform_protocol_create(kaa_platform_protocol_t **platform_prot
...
+    }
 
     uint16_t total_services_count = services_count + 1 /* Meta extension */;
 
-    error_code = kaa_platform_message_header_write(writer, KAA_PLATFORM_PROTOCOL_ID, KAA_PLATFORM_PROTOCOL_VERSION);
+    error_code = kaa_platform_message_header_write(writer,
+            KAA_PLATFORM_PROTOCOL_ID, KAA_PLATFORM_PROTOCOL_VERSION);
     if (error_code) {
         KAA_LOG_ERROR(self->logger, error_code, "Failed to write the client sync header");
         return error_code;
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: `writer` has leaked.

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.h [hunk 3]
```diff
@@ -55,19 +55,40 @@ typedef struct {
...
+ * the buffer.
  *
- * @param[in]  self        Pointer to a @link kaa_platform_protocol_t @endlink instance.
- * @param[in]  info        Pointer to a @link kaa_serialize_info_t @endlink instance.
- * @param[out] buffer      The buffer with serialized data.
- * @param[out] buffer_size The buffer's actual size.
+ * If buffer size is not enough, @c KAA_ERR_BUFFER_IS_NOT_ENOUGH is
+ * returned. You should reallocate buffer and repeat the call.
+ *
+ * @param[in]  self           Pointer to a @link kaa_platform_protocol_t @endlink instance.
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.h]: Use `@ref` instead https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdref

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 4]
```diff
@@ -153,126 +164,159 @@ static kaa_error_t kaa_client_sync_get_size(kaa_platform_protocol_t *self,
...
 }
 
-
-
-static kaa_error_t kaa_client_sync_serialize(kaa_platform_protocol_t *self
-                                           , const kaa_extension_id services[]
-                                           , size_t services_count
-                                           , char* buffer
-                                           , size_t *size)
+// TODO: make buffer uint8_t
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: There are a lot of places where char has to be replaced with uint8_t. I think it is rational to create issues about that.

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 5]
```diff
@@ -153,126 +164,159 @@ static kaa_error_t kaa_client_sync_get_size(kaa_platform_protocol_t *self,
...
     if (error_code) {
         KAA_LOG_ERROR(self->logger, error_code, "Failed to write the client sync header");
-        return error_code;
+        goto fail;
     }
-    char *extension_count_p = writer->current;
+
+    uint16_t *extension_count_p = (uint16_t *)writer->current;
     writer->current += KAA_PROTOCOL_EXTENSIONS_COUNT_SIZE;
+    // TODO: assert KAA_PROTOCOL_EXTENSIONS_COUNT_SIZE == sizeof(uint16_t)
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: KAA-982 ?

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.h [hunk 6]
```diff
@@ -36,20 +36,15 @@
 extern "C" {
 #endif
 
+#ifndef KAA_PLATFORM_PRTOCOL_T
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.h]: Already defined in 25 line

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.h [hunk 7]
```diff
@@ -58,37 +53,34 @@ typedef struct {
...
- * perfectly acceptable to return a pointer to a previously allocated buffer (even on the stack) if its size is sufficient.
+ * @brief Constructs a sync request for the specified list of services
+ * based on the current state of Kaa context and serializes it into
+ * the buffer returned by the allocator function.
  *
- * @param[in]  self              Pointer to a @link kaa_platform_protocol_t @endlink instance.
- * @param[in]  info              Pointer to a @link kaa_serialize_info_t @endlink instance.
- * @param[out] buffer            The buffer with serialized data.
- * @param[out] buffer_size       The buffer's actual size.
+ * @param[in]  self        Pointer to a @link kaa_platform_protocol_t @endlink instance.
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.h]: Use `@ref`

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.h [hunk 8]
```diff
@@ -58,37 +53,34 @@ typedef struct {
...
+        char **buffer,
+        size_t *buffer_size);
 
 /**
  * @brief Processes downstream data received from Operations server.
  *
- * @param[in] self              Pointer to a @link kaa_platform_protocol_t @endlink instance.
- * @param[in] buffer            Pointer to a data buffer for processing received from Operations server.
- * @param[in] buffer_size       Size of @c buffer.
+ * @param[in] self        Pointer to a @link kaa_platform_protocol_t @endlink instance.
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.h]: Use `@ref`

**Comment by forGGe** [comment]
Reviewed

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c (hunk 4 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: Don't agree. We can't change _all_ buffer in one go, so we should do this gradually along other tasks.

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c (hunk 5 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: That should be a static assert, not a dynamic one.

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c [hunk 9]
```diff
@@ -1118,7 +1117,7 @@ kaa_error_t kaa_tcp_channel_authorize(kaa_tcp_channel_t *self)
     kaatcp_error_t kaatcp_error_code =
             kaatcp_fill_connect_message(KAA_TCP_CHANNEL_MAX_TIMEOUT
                                       , KAA_PLATFORM_PROTOCOL_ID
-                                      , sync_buffer
+                                      , (char *)sync_buffer
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c]: Remove cast

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c [hunk 10]
```diff
@@ -1520,7 +1517,7 @@ kaa_error_t kaa_tcp_channel_write_pending_services(kaa_tcp_channel_t *self
         return error_code;
     }
 
-    kaatcp_error_t parser_error_code = kaatcp_fill_kaasync_message(sync_buffer
+    kaatcp_error_t parser_error_code = kaatcp_fill_kaasync_message((char *)sync_buffer
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c]: Remove cast

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c (hunk 5 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: Changed comment to reflect that

**Comment by rasendubi** [comment]
@forGGe, I've addressed all comments, re-review, please.

**Comment by forGGe** [comment]
+

**Pull request merged**